docs/design/apiserver_storage.md

```diff
@@ -0,0 +1,153 @@
+
+- CPU and memory efficient: The storage layer should be designed to support object indexing and watch indexing. Storage layer should be designed to hide all storage related optimization details. The storage should be the single place to optimize server storage. If an index is built, users should expect to select objects in reasonable amount of time (for example < O(logN) for list). To optimize caching size of pod objects, we should do all the work in storage system without touching the pod object defined at API layer.
+
+- easy to troubleshoot: Users should be able to identify easily if the storage layer is the current bottleneck of their cluster and why. We should expose enough metrics and write understandable docs to tell users how to do the troubleshooting.
+
+- easy to support multiple generic storage backends: The storage layer should have a clear separation between operations on Kubernetes objects and generic key-value pairs. To implement a new key-value or database backend, developers should not have to write duplicated code to handle Kubernetes logic. Ideally, they do not even need to know anything about Kubernetes at all.
+
+- backward compatible: Any changes should not introduce backward compatibility issues across at least last two minor leases from the end-user perspective. Using the improved storage system with existing etcd3 backend should just work.
+
+## Direction
+
+We could evolve the current storage layer towards the goal we listed above. It is doable, but it might involve significant interface and implement changes. Also it requires continuous changes of the consumers of the current storage API. During this process, we also need to ensure everything works perfectly. There is no alpha, or beta since we will be changing the single critical path. This is risky, time consuming, and hard to revert.
+
+The solution we proposed here is to start a storage layer from scratch. As Daniel Smith [mentioned](https://github.com/kubernetes/kubernetes/pull/28508#issuecomment-232508836)
+
+> Just to (probably unnecessarily) reiterate my position, I think we're going to need side-by-side implementations of both old and new storage styles, so we shouldn't end up locked into whatever we do now.
+
+We also believe we should start a side-by-side storage implementation to reach the goal we listed. That way, we won’t be affected by any existing issues, and it gives us the opportunities to rethink the design from the ground up, to achieve the goals more clearly. The only risk is that the effort will be completely wasted, if we cannot reach the goals.
+
+## Proposed Storage API
+
+```go
+
+type Storage interface {
+	// Wait waits until the storage has seen any version >= given version.
+	Wait(ctx context.Context, version uint64)
```

Comment:
Why do you assume I'm referring to etcd? What is too specific here? How could you implement MVCC transaction without numeric index?